CosmosDB--Added specs for Get Metrics, Get Usages, Get MetricDefinitions

molly-moen opened the pull request:
Added the specifications for cosmos db paths for get metrics, get usages and get metric definitions. Also includes example files for these types of requests.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [y ] The title of the PR is clear and informative.
- [y] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [y ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [y ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ y] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [y ] My spec meets the review criteria:
  - [ y] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ y] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ y] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

molly-moen committed:
metrics-documentation

started work on metrics documentation

molly-moen committed:
Added get metrics, get usages, get metric definitions

Added specifications for get metrics, get usages and get metric definitions paths

molly-moen committed:
Fixed indentation and some descriptions

Fixed indentation and some descriptions

molly-moen committed:
Changed Get to List

Validation requires array result to have operation id including _List

olydis commented:
@simongdavies new APIs

olydis commented:
@ravbhatnagar

ravbhatnagar commented:
Looks good

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -437,26 +437,432 @@
+            "$ref": "#/parameters/filterParameter"
+          }
+        ],
+        "responses": {
```
looking at `MetricListResult`, it looks to me as if this operation should be marked as `x-ms-pageable` (with `nextLinkName` set to `null`)

molly-moen replied:
It's never going to return it as multiple pages

olydis replied:
that's why I suggested setting `nextLinkName` set to `null` 😉 it still improves generated code, since not a wrapper object with an array property is returned, but instead just the array itself. Of course, ignore my suggestion if you think there could be further properties appearing as siblings of `values` - since then, you would of course want users to see the "wrapper object"

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -437,26 +437,432 @@
+            "$ref": "#/parameters/filterParameter"
+          }
+        ],
+        "responses": {
```
same as above 🙂

olydis replied:
...and next 3 times :D

molly-moen replied:
same--not going to be multiple pages

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -437,26 +437,432 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
okay, same here, please review all added methods 😉

molly-moen replied:
same--not going to be multiple pages

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -437,26 +437,432 @@
+        "x-ms-examples": {
+          "CosmosDBDatabaseGetUsages": { "$ref": "./examples/CosmosDBDatabaseGetUsages.json" }
+        },
+        "description": "Retrieves the usages (most recent data) for the given database account.",
```
description identical to `DatabaseAccounts_ListUsages`

molly-moen replied:
fixed

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -437,26 +437,432 @@
+        "x-ms-examples": {
+          "CosmosDBCollectionGetUsages": { "$ref": "./examples/CosmosDBCollectionGetUsages.json" }
+        },
+        "description": "Retrieves the usages (most recent storage data) for the given database account.",
```
description identical to `DatabaseAccounts_ListUsages`

molly-moen replied:
fixed

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -921,6 +1327,277 @@
+      "description": "The response to a list usage request."
+    },
+    "Usage": {
+      "properties": {
```
Can the properties in here be marked as `readOnly`? So far, they're only used in responses it seems. 🙂

molly-moen replied:
fixed

olydis commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/2015-04-08/cosmos-db.json:
```diff
@@ -921,6 +1327,277 @@
+    "Usage": {
+      "properties": {
+        "unit": {
+          "type": "string",
```
please extract `UnitType` into its own definition section and just reference it here => deduplication (generally applies to anything that is duplicated across a spec)

molly-moen replied:
fixed

olydis commented:
@molly-moen Please also take a look at our [model validation](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/309633756#L702) - it found a load of mismatches between the examples you provided and the specification.

molly-moen committed:
Fixed model validation and addressed comments

Fixed readonly/required attributes, extracted out repeated text, fixed some casing and descriptions

molly-moen committed:
Changed List Metrics to be x-ms-pageable

Changed List Metrics to be x-ms-pageable with next link name of null
